### PR TITLE
WIP/RFC: snap pkg docs & deps & internals

### PR DESCRIPTION
**Pull request by ppalucki**
Fixes:
- https://godoc.org/github.com/intelsdi-x/snap missing descriptions of packages

Summary of changes:
- pkg descriptions (`doc.go` files)
- dependencies (as ascii graphs) - see [here](https://github.com/ppalucki/snap/blob/pkgdocs/docs/INTERNALS.md)
- internal overview of components relationship (INTERNALS.md)

TODO:
- [ ] describe mgmt & control package
- [ ] extend description of each/some packages
- [ ] missing pkg dependency diagrams mgmt (readable version of [this](https://godoc.org/github.com/intelsdi-x/snap?import-graph&hide=2) )

@intelsdi-x/snap-maintainers

Results is:
```
➜ go list -f "{{.ImportPath}}: {{.Doc}}" ./...
github.com/intelsdi-x/snap:
github.com/intelsdi-x/snap/cmd/snapctl:
github.com/intelsdi-x/snap/control: Package control PluginManger manages loading, unloading and swapping of plugins.
github.com/intelsdi-x/snap/control/plugin: Package plugin is a layer that connects external plugins with snap core (plugin metadata, sessions and own PluginMetricType) as well as Collector, Publisher and Processor abstractions.
github.com/intelsdi-x/snap/control/plugin/client: Package client provides layer how snap core communicate with external plugins.
github.com/intelsdi-x/snap/control/plugin/cpolicy: Package cpolicy defines "configuration policy" ConfigPolicy based on rules, that specifiy what configuration is required by plugin.
github.com/intelsdi-x/snap/control/plugin/encoding:
github.com/intelsdi-x/snap/control/plugin/encrypter:
github.com/intelsdi-x/snap/control/strategy: Package strategy provides basic interfaces for routing to available plugins and caching metric data.
github.com/intelsdi-x/snap/core: Package core defines base interfaces like: Metric, Plugin, Task & Workflow.
github.com/intelsdi-x/snap/core/cdata: Package cdata defines configuration structs.
github.com/intelsdi-x/snap/core/control_event: Package control_event contains controller events.
github.com/intelsdi-x/snap/core/ctypes: Package ctypes defines differnt kinds of ConfigTypes like Int, Float, Bool, Str.
github.com/intelsdi-x/snap/core/scheduler_event: Package scheduler_event contains scheduler events (structs) like TaskCreated, TaskStarted, ...
github.com/intelsdi-x/snap/core/serror: Package serror defines custom error SnapError with additional fields.
github.com/intelsdi-x/snap/mgmt/rest:
github.com/intelsdi-x/snap/mgmt/rest/client:
github.com/intelsdi-x/snap/mgmt/rest/rbody:
github.com/intelsdi-x/snap/mgmt/rest/request:
github.com/intelsdi-x/snap/mgmt/tribe:
github.com/intelsdi-x/snap/mgmt/tribe/agreement:
github.com/intelsdi-x/snap/mgmt/tribe/worker:
github.com/intelsdi-x/snap/pkg/aci: Package with helper functions to handle extraction and validation of ACI Images.
github.com/intelsdi-x/snap/pkg/chrono:
github.com/intelsdi-x/snap/pkg/ctree: Package ctree implements ConfigTree that can store Node under the given keys ([]string).
github.com/intelsdi-x/snap/pkg/globalconfig:
github.com/intelsdi-x/snap/pkg/promise:
github.com/intelsdi-x/snap/pkg/psigning:
github.com/intelsdi-x/snap/pkg/schedule:
github.com/intelsdi-x/snap/plugin/collector/snap-collector-mock1:
github.com/intelsdi-x/snap/plugin/collector/snap-collector-mock1/mock:
github.com/intelsdi-x/snap/plugin/collector/snap-collector-mock2:
github.com/intelsdi-x/snap/plugin/collector/snap-collector-mock2/mock:
github.com/intelsdi-x/snap/plugin/helper: Package helper The test helper for testing plugins
github.com/intelsdi-x/snap/plugin/processor/snap-processor-passthru:
github.com/intelsdi-x/snap/plugin/processor/snap-processor-passthru/passthru:
github.com/intelsdi-x/snap/plugin/publisher/snap-publisher-file:
github.com/intelsdi-x/snap/plugin/publisher/snap-publisher-file/file:
github.com/intelsdi-x/snap/scheduler: Package scheduler implements scheduler for workflow.
github.com/intelsdi-x/snap/scheduler/wmap: Package wmap provides implementation of Workflow.

```

### CODE REVIEW

File: core/task.go [hunk 1]
```diff
@@ -91,6 +91,15 @@ type Task interface {
...
 
+type A int
+
+type B struct {
+	foo int
+	bar int
+}
+
+type x func(int) x
+
```
Comment (danielscottt) [inline core/task.go]: Hey @ppalucki 

What are these for?

File: docs/INTERNALS.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+       | Scheduler |
+       +----+------+
+            |
+  run as Job|
+            |                                    is a      +-----------------+  produce  +--------+
+        +---v--+  defined as                     +---------> CollectorPlugin +-----------+ Metric |
+        | Task +---------+                       |         +-----------------+           +---+----+
+        +-+----+         |                       |                                           |
+          |       +------v---+  build from +-----+--+ is a +-----------------+               |
+accordning|       | Workflow +-------------> Plugin +------> ProccesorPlugin <---------------+
```
Comment (danielscottt) [inline docs/INTERNALS.md]: s/accordning/according

File: docs/INTERNALS.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+          |       +----------+             +-^---+--+      +-----------------+               |
+  +-------v---+                              |   |                |as bytes          as bytes|
+  | Schedule  |                              |   | is a    +------v----------+               |
+  +-----------+                              |   +---------> PublisherPlugin <---------------+
+                                             |             +------+----------+
+                                       manage|                    |
+                                             |                    |              +------------+
+      +---------------+ thorugh +---------+  |                    |store in      |  External  |
```
Comment (danielscottt) [inline docs/INTERNALS.md]: s/thorugh/through

File: core/task.go (hunk 1 above)
Comment (ppalucki) [inline core/task.go]: argh, I was trying to explain my coworker, how the TaskOption (undoing) works - the recursive func type - and forget to remove example - I'll delete it 
ps. is "undoing" feature used somewhere ? I couldn't find any usage - return value from calling is always ignored - or I'm missing something

Comment (danielscottt) [inline core/task.go]: It is not! You'll find that the later iterations of this pattern don't use
it, e.g. options on plugin meta.

**Comment by ppalucki** [comment]
ps. I've used [prdeps](http://dave.cheney.net/2016/02/29/investigate-your-package-dependencies-with-prdeps-and-unix) to find dependencies and vim/drawIt if someone is curious

**Comment by candysmurf** [comment]
Last time team didn't prefer to use doc.go while I worked on docs. I'm not sure if the decision is changed now.

**Comment by ppalucki** [comment]
Closing this one in favor of my [working branch](https://github.com/ppalucki/snap/tree/serenity) for prototyping complex processors workflows - I'll continue documenting internals there and when sure I'm know what I'm talking about (I was wrong about wmap and scheduler packages), push as new PR.

**Pull request closed without merging**